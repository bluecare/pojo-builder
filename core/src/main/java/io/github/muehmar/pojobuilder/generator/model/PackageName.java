package io.github.muehmar.pojobuilder.generator.model;

import lombok.Value;

@Value
public class PackageName {
  String pkg;

  private PackageName(String pkg) {
    this.pkg = pkg;
  }

  public static PackageName fromString(String pkg) {
    return new PackageName(pkg);
  }

  public static PackageName javaLang() {
    return fromString("java.lang");
  }

  public static PackageName javaUtil() {
    return fromString("java.util");
  }

  public Name qualifiedName(Name className) {
    return className.prefix(".").prefix(pkg);
  }

  public String asString() {
    return pkg;
  }

  @Override
  public String toString() {
    return asString();
  }
}
