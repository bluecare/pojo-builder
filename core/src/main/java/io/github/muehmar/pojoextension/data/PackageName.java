package io.github.muehmar.pojoextension.data;

import java.util.Objects;

public class PackageName {
  private final String pkg;

  private PackageName(String pkg) {
    this.pkg = pkg;
  }

  public static PackageName fromString(String pkg) {
    return new PackageName(pkg);
  }

  public static PackageName javaLang() {
    return fromString("java.lang");
  }

  public static PackageName javaUtil() {
    return fromString("java.util");
  }

  public String asString() {
    return pkg;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    PackageName packageName = (PackageName) o;
    return Objects.equals(pkg, packageName.pkg);
  }

  @Override
  public int hashCode() {
    return Objects.hash(pkg);
  }

  @Override
  public String toString() {
    return "Package{" + "pkg='" + pkg + '\'' + '}';
  }
}
