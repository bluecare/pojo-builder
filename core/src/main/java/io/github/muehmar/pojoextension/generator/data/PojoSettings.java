package io.github.muehmar.pojoextension.generator.data;

import io.github.muehmar.pojoextension.annotations.PojoExtension;
import java.util.Optional;

@PojoExtension
@SuppressWarnings("java:S2160")
public class PojoSettings extends PojoSettingsExtension {
  private final boolean disableSafeBuilder;
  private final Optional<Name> extensionName;

  public PojoSettings(boolean disableSafeBuilder, Optional<Name> extensionName) {
    this.disableSafeBuilder = disableSafeBuilder;
    this.extensionName = extensionName;
  }

  public static PojoSettings defaultSettings() {
    return new PojoSettings(false, Optional.empty());
  }

  public boolean isDisableSafeBuilder() {
    return disableSafeBuilder;
  }

  public boolean isEnableSafeBuilder() {
    return !isDisableSafeBuilder();
  }

  public Name qualifiedExtensionName(Pojo pojo) {
    return pojo.getPkg().qualifiedName(extensionName(pojo));
  }

  public Optional<Name> getExtensionName() {
    return extensionName;
  }

  public Name extensionName(Pojo pojo) {
    return extensionName.orElseGet(
        () -> pojo.getName().map(n -> n.replace(".", "")).append("Extension"));
  }
}
