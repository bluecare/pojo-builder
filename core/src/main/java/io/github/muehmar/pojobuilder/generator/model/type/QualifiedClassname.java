package io.github.muehmar.pojobuilder.generator.model.type;

import io.github.muehmar.pojobuilder.generator.model.Name;
import io.github.muehmar.pojobuilder.generator.model.PackageName;
import lombok.Value;

@Value
public class QualifiedClassname {
  Classname classname;
  PackageName pkg;

  public Name getName() {
    return classname.asName();
  }

  public Name getSimpleName() {
    return classname.getSimpleName();
  }

  public String asString() {
    return String.format("%s.%s", pkg.asString(), classname.asString());
  }

  @Override
  public String toString() {
    return asString();
  }
}
