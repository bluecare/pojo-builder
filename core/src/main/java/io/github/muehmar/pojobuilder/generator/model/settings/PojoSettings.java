package io.github.muehmar.pojobuilder.generator.model.settings;

import static java.util.Optional.empty;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojobuilder.annotations.FullBuilderFieldOrder;
import io.github.muehmar.pojobuilder.annotations.OptionalDetection;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import io.github.muehmar.pojobuilder.generator.model.ClassAccessLevelModifier;
import io.github.muehmar.pojobuilder.generator.model.Name;
import io.github.muehmar.pojobuilder.generator.model.Pojo;
import java.util.Optional;
import lombok.Value;
import lombok.With;

@Value
@With
@PojoBuilder
public class PojoSettings {
  private static final Name CLASS_NAME_PLACEHOLDER = Name.fromString("{CLASSNAME}");
  public static final Name BUILDER_CLASS_POSTFIX = Name.fromString("Builder");
  PList<OptionalDetection> optionalDetections;
  Optional<Name> builderName;
  Optional<Name> builderSetMethodPrefix;
  ClassAccessLevelModifier builderAccessLevel;
  boolean standardBuilderEnabled;
  boolean fullBuilderEnabled;
  FullBuilderFieldOrder fullBuilderFieldOrder;

  public static PojoSettings defaultSettings() {
    return PojoSettingsBuilder.create()
        .optionalDetections(
            PList.of(OptionalDetection.OPTIONAL_CLASS, OptionalDetection.NULLABLE_ANNOTATION))
        .builderAccessLevel(ClassAccessLevelModifier.PUBLIC)
        .standardBuilderEnabled(true)
        .fullBuilderEnabled(true)
        .fullBuilderFieldOrder(FullBuilderFieldOrder.REQUIRED_FIELDS_FIRST)
        .andAllOptionals()
        .builderName(Optional.of(CLASS_NAME_PLACEHOLDER.append(BUILDER_CLASS_POSTFIX)))
        .builderSetMethodPrefix(empty())
        .build();
  }

  public Name qualifiedBuilderName(Pojo pojo) {
    return pojo.getPackage().qualifiedName(builderName(pojo));
  }

  public Name builderName(Pojo pojo) {
    return getNameOrAppend(builderName, BUILDER_CLASS_POSTFIX, pojo);
  }

  private Name getNameOrAppend(Optional<Name> name, Name postfix, Pojo pojo) {
    return name.map(n -> n.replace(CLASS_NAME_PLACEHOLDER, getClassName(pojo)))
        .orElseGet(() -> getClassName(pojo).append(postfix));
  }

  private Name getClassName(Pojo pojo) {
    return pojo.getName().map(n -> n.replace(".", ""));
  }

  public PojoSettings withBuilderNameOpt(Name builderName) {
    return this.withBuilderName(Optional.of(builderName));
  }

  public PojoSettings withBuilderSetMethodPrefixOpt(Name builderSetMethodPrefix) {
    return this.withBuilderSetMethodPrefix(Optional.of(builderSetMethodPrefix));
  }

  public PojoSettings overrideOptionalDetection(
      Optional<PList<OptionalDetection>> optionalDetections) {
    return optionalDetections.map(this::withOptionalDetections).orElse(this);
  }

  public PojoSettings overrideBuilderName(Optional<Name> builderName) {
    return builderName.map(this::withBuilderNameOpt).orElse(this);
  }

  public PojoSettings overrideBuilderSetMethodPrefix(Optional<Name> builderSetMethodPrefix) {
    return builderSetMethodPrefix
        .map(ignore -> this.withBuilderSetMethodPrefix(builderSetMethodPrefix))
        .orElse(this);
  }

  public PojoSettings overrideBuilderAccessLevel(
      Optional<ClassAccessLevelModifier> classAccessLevelModifier) {
    return classAccessLevelModifier.map(this::withBuilderAccessLevel).orElse(this);
  }
}
