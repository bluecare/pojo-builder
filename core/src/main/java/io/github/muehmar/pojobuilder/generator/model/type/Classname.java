package io.github.muehmar.pojobuilder.generator.model.type;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojobuilder.generator.model.Name;
import lombok.EqualsAndHashCode;

/**
 * Name of a class. Could also be an inner class (or static nested class) where it is referenced via
 * its enclosing class(es) like OuterClass.InnerClass.
 */
@EqualsAndHashCode
public class Classname {
  private final PList<Name> outerClassNames;
  private final Name simpleName;

  private Classname(PList<Name> outerClassNames, Name simpleName) {
    this.outerClassNames = outerClassNames;
    this.simpleName = simpleName;
  }

  public static Classname fromString(String fullClassName) {
    final PList<Name> classesReversed =
        PList.fromArray(fullClassName.split("\\.")).map(Name::fromString).reverse();
    if (classesReversed.isEmpty()) {
      throw new IllegalArgumentException("Invalid class name " + fullClassName);
    }

    final Name className = classesReversed.head();
    return new Classname(classesReversed.drop(1).reverse(), className);
  }

  /**
   * Returns the top level class for this class name. This is either the class itself if it's a top
   * level class or the outermost class for an inner class.
   */
  public Name getTopLevelClass() {
    return outerClassNames.headOption().orElse(simpleName);
  }

  public Name getSimpleName() {
    return simpleName;
  }

  public Name asName() {
    return Name.fromString(asString());
  }

  public String asString() {
    return outerClassNames.add(simpleName).mkString(".");
  }

  @Override
  public String toString() {
    return asString();
  }
}
