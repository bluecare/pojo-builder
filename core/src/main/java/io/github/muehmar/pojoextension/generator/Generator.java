package io.github.muehmar.pojoextension.generator;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojoextension.generator.writer.Writer;
import java.util.function.BiPredicate;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.UnaryOperator;

public interface Generator<A, B> {
  Writer generate(A data, B settings, Writer writer);

  static <A, B> Generator<A, B> ofWriterFunction(UnaryOperator<Writer> f) {
    return (data, settings, writer) -> f.apply(writer);
  }

  static <A, B> Generator<A, B> emptyGen() {
    return (data, settings, writer) -> writer;
  }

  default Generator<A, B> appendNewLine() {
    return append((UnaryOperator<Writer>) Writer::println);
  }

  default Generator<A, B> append(Generator<A, B> next) {
    final Generator<A, B> self = this;
    return (data, settings, writer) ->
        next.generate(data, settings, self.generate(data, settings, writer));
  }

  default Generator<A, B> append(Generator<A, B> next, int tabs) {
    final Generator<A, B> self = this;
    return (data, settings, writer) ->
        self.generate(data, settings, writer)
            .append(tabs, next.generate(data, settings, writer.empty()));
  }

  default Generator<A, B> append(UnaryOperator<Writer> next) {
    return append((data, settings, writer) -> next.apply(writer));
  }

  default <C> Generator<A, B> append(Generator<C, B> gen, Function<A, C> f) {
    final Generator<A, B> self = this;
    return (data, settings, writer) ->
        gen.generate(f.apply(data), settings, self.generate(data, settings, writer));
  }

  default <C> Generator<A, B> appendList(Generator<C, B> gen, Function<A, PList<C>> f) {
    final Generator<A, B> self = this;
    return (data, settings, writer) ->
        f.apply(data)
            .foldLeft(self, (g, e) -> g.append(gen, ignore -> e))
            .generate(data, settings, writer);
  }

  default Generator<A, B> appendConditionally(BiPredicate<A, B> predicate, Generator<A, B> append) {
    final Generator<A, B> self = this;
    return (data, settings, writer) -> {
      if (predicate.negate().test(data, settings)) {
        return self.generate(data, settings, writer);
      }
      return append(append).generate(data, settings, writer);
    };
  }

  default Generator<A, B> appendConditionally(Predicate<A> predicate, Generator<A, B> append) {
    return appendConditionally((data, settings) -> predicate.test(data), append);
  }
}
