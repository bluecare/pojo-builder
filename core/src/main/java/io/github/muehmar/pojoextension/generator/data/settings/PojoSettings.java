package io.github.muehmar.pojoextension.generator.data.settings;

import static io.github.muehmar.pojoextension.generator.data.settings.Ability.ENABLED;
import static io.github.muehmar.pojoextension.generator.data.settings.ExtensionUsage.INHERITED;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojoextension.annotations.OptionalDetection;
import io.github.muehmar.pojoextension.annotations.PojoExtension;
import io.github.muehmar.pojoextension.generator.data.Name;
import io.github.muehmar.pojoextension.generator.data.Pojo;
import java.util.Optional;

@PojoExtension
@SuppressWarnings("java:S2160")
public class PojoSettings extends PojoSettingsBase {
  private static final Name CLASS_NAME_PLACEHOLDER = Name.fromString("{CLASSNAME}");
  public static final Name BUILDER_CLASS_POSTFIX = Name.fromString("Builder");
  public static final Name EXTENSION_IFC_POSTFIX = Name.fromString("Extension");
  public static final Name BASE_CLASS_POSTFIX = Name.fromString("Base");
  private final PList<OptionalDetection> optionalDetections;
  private final ExtensionUsage extensionUsage;
  private final Optional<Name> extensionName;
  private final Optional<Name> builderName;
  private final Optional<Name> baseClassName;
  private final Ability safeBuilderAbility;
  private final Ability equalsHashCodeAbility;
  private final Ability toStringAbility;
  private final Ability withersAbility;
  private final Ability mappersAbility;
  private final Ability baseClassAbility;

  PojoSettings(
      PList<OptionalDetection> optionalDetections,
      ExtensionUsage extensionUsage,
      Optional<Name> extensionName,
      Optional<Name> builderName,
      Optional<Name> baseClassName,
      Ability safeBuilderAbility,
      Ability equalsHashCodeAbility,
      Ability toStringAbility,
      Ability withAbility,
      Ability mapAbility,
      Ability baseClassAbility) {
    this.optionalDetections = optionalDetections;
    this.extensionUsage = extensionUsage;
    this.extensionName = extensionName;
    this.builderName = builderName;
    this.baseClassName = baseClassName;
    this.safeBuilderAbility = safeBuilderAbility;
    this.equalsHashCodeAbility = equalsHashCodeAbility;
    this.toStringAbility = toStringAbility;
    this.withersAbility = withAbility;
    this.mappersAbility = mapAbility;
    this.baseClassAbility = baseClassAbility;
  }

  public static PojoSettings defaultSettings() {
    return PojoSettingsBuilder.create()
        .setOptionalDetections(
            PList.of(OptionalDetection.OPTIONAL_CLASS, OptionalDetection.NULLABLE_ANNOTATION))
        .setExtensionUsage(INHERITED)
        .setSafeBuilderAbility(ENABLED)
        .setEqualsHashCodeAbility(ENABLED)
        .setToStringAbility(ENABLED)
        .setWithersAbility(ENABLED)
        .setMappersAbility(ENABLED)
        .setBaseClassAbility(ENABLED)
        .andAllOptionals()
        .setExtensionName(Optional.of(CLASS_NAME_PLACEHOLDER.append(EXTENSION_IFC_POSTFIX)))
        .setBuilderName(Optional.of(CLASS_NAME_PLACEHOLDER.append(BUILDER_CLASS_POSTFIX)))
        .setBaseClassName(Optional.of(CLASS_NAME_PLACEHOLDER.append(BASE_CLASS_POSTFIX)))
        .build();
  }

  public PList<OptionalDetection> getOptionalDetections() {
    return optionalDetections;
  }

  public ExtensionUsage getExtensionUsage() {
    return extensionUsage;
  }

  public Optional<Name> getExtensionName() {
    return extensionName;
  }

  public Optional<Name> getBuilderName() {
    return builderName;
  }

  public Optional<Name> getBaseClassName() {
    return baseClassName;
  }

  public Ability getSafeBuilderAbility() {
    return safeBuilderAbility;
  }

  public Ability getEqualsHashCodeAbility() {
    return equalsHashCodeAbility;
  }

  public Ability getToStringAbility() {
    return toStringAbility;
  }

  public Ability getWithersAbility() {
    return withersAbility;
  }

  public Ability getMappersAbility() {
    return mappersAbility;
  }

  public Ability getBaseClassAbility() {
    return baseClassAbility;
  }

  public Name qualifiedExtensionName(Pojo pojo) {
    return pojo.getPackage().qualifiedName(extensionName(pojo));
  }

  public Name extensionName(Pojo pojo) {
    return getNameOrAppend(extensionName, EXTENSION_IFC_POSTFIX, pojo);
  }

  public Name qualifiedBuilderName(Pojo pojo) {
    return pojo.getPackage().qualifiedName(builderName(pojo));
  }

  public Name builderName(Pojo pojo) {
    return getNameOrAppend(builderName, BUILDER_CLASS_POSTFIX, pojo);
  }

  public Name qualifiedBaseClassName(Pojo pojo) {
    return pojo.getPackage().qualifiedName(baseClassName(pojo));
  }

  public Name baseClassName(Pojo pojo) {
    return getNameOrAppend(baseClassName, BASE_CLASS_POSTFIX, pojo);
  }

  private Name getNameOrAppend(Optional<Name> name, Name postfix, Pojo pojo) {
    return name.map(n -> n.replace(CLASS_NAME_PLACEHOLDER, getClassName(pojo)))
        .orElseGet(() -> getClassName(pojo).append(postfix));
  }

  private Name getClassName(Pojo pojo) {
    return pojo.getName().map(n -> n.replace(".", ""));
  }

  public boolean createDiscreteBuilder() {
    return safeBuilderAbility.isEnabled();
  }

  public boolean createExtension() {
    return equalsHashCodeAbility.isEnabled()
        || toStringAbility.isEnabled()
        || withersAbility.isEnabled()
        || mappersAbility.isEnabled();
  }

  public boolean createBaseClass() {
    return (equalsHashCodeAbility.isEnabled() || toStringAbility.isEnabled())
        && baseClassAbility.isEnabled();
  }
}
