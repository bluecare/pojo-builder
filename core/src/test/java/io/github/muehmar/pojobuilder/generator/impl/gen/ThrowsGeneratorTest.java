package io.github.muehmar.pojobuilder.generator.impl.gen;

import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;
import static io.github.muehmar.pojobuilder.generator.impl.gen.ThrowsGenerator.throwsGenerator;
import static io.github.muehmar.pojobuilder.snapshottesting.SnapshotUtil.writerSnapshot;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import ch.bluecare.commons.data.PList;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import io.github.muehmar.pojobuilder.generator.model.PackageName;
import io.github.muehmar.pojobuilder.generator.model.settings.PojoSettings;
import io.github.muehmar.pojobuilder.generator.model.type.Classname;
import io.github.muehmar.pojobuilder.generator.model.type.QualifiedClassname;
import io.github.muehmar.pojobuilder.snapshottesting.SnapshotTest;
import org.junit.jupiter.api.Test;

@SnapshotTest
class ThrowsGeneratorTest {
  private Expect expect;

  @Test
  @SnapshotName("twoExceptions")
  void generate_when_twoExceptions_then_matchSnapshot() {
    final Generator<PList<QualifiedClassname>, PojoSettings> generator = throwsGenerator();

    final PList<QualifiedClassname> exceptions =
        PList.of(
            new QualifiedClassname(
                Classname.fromString("SomeException"),
                PackageName.fromString("io.github.muehmar.exception")),
            new QualifiedClassname(
                Classname.fromString("IllegalArgumentException"), PackageName.javaLang()));

    final Writer writer =
        generator.generate(exceptions, PojoSettings.defaultSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }
}
