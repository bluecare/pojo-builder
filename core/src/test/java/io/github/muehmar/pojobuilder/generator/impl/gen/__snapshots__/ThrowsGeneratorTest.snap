twoExceptions=[
io.github.muehmar.exception.SomeException
java.lang.IllegalArgumentException

SomeException, IllegalArgumentException
]